### PR TITLE
Forbid removing the last admin user

### PR DESCRIPTION
**Pull request by yp05327**

### CODE REVIEW

File: models/user/user.go [hunk 1]
```diff
@@ -1191,6 +1191,11 @@ func GetAdminUser(ctx context.Context) (*User, error) {
 	return &admin, nil
 }
 
+// GetAdminUserCount returns the count of all administrator
+func GetAdminUserCount(ctx context.Context) (int64, error) {
```
Comment (wolfogre) [inline models/user/user.go]: Maybe `CountAdminUsers` or reuse `CountUsers` with additional conditions in `CountUserFilter`.

**Comment by lunny** [comment]
I think it has been implemented so that the admin cannot delete himself.

File: models/user/user.go (hunk 1 above)
Comment (delvh) [reply models/user/user.go]: `db.Count`?

**Comment by yp05327** [comment]
Yes. I did this in Gitea Cloud Instance.
Then I can not access the admin page again because there are no admin user in that instance.

File: models/user/user.go (hunk 1 above)
Comment (yp05327) [reply models/user/user.go]: `CountUsers` would not return errors. Will it cause issues here?

ps: I added print log in it, so when there's an errors occurred, we can trace it.

**Comment by yp05327** [comment]
> I think it has been implemented so that the admin cannot delete himself.

Maybe this only implemented in admin settings, as I successfully delete the last admin himself in user settings. 😅

**Comment by yp05327** [comment]
↑ I will also fix it in this PR

**Comment by lunny** [comment]
lint error is related.

File: routers/web/user/setting/account.go [hunk 2]
```diff
@@ -246,6 +246,13 @@ func DeleteAccount(ctx *context.Context) {
 		return
 	}
 
+	// admin should not delete themself
+	if ctx.Doer.IsAdmin {
+		ctx.Flash.Error(ctx.Tr("admin.users.cannot_delete_self"))
```
Comment (wolfogre) [inline routers/web/user/setting/account.go]: Just a small suggestion.

`admin.users.cannot_delete_self` ("You cannot delete yourself") is used to inform the user that they cannot delete themselves on the admin page. However, this is the account page, and the user is indeed deleting themselves. So, maybe the error message could be something like "You cannot delete yourself when you are an admin. Please remove your admin privileges first."

**Review by wolfogre — APPROVED** [review]
LGTM.

File: routers/web/user/setting/account.go (hunk 2 above)
Comment (yp05327) [reply routers/web/user/setting/account.go]: Done.

File: models/user/user.go [hunk 3]
```diff
@@ -716,13 +725,25 @@ func CountUsers(ctx context.Context, opts *CountUserFilter) int64 {
 }
 
 func countUsers(ctx context.Context, opts *CountUserFilter) int64 {
-	sess := db.GetEngine(ctx).Where(builder.Eq{"type": "0"})
+	sess := db.GetEngine(ctx)
+	cond := builder.NewCond()
+	cond = cond.And(builder.Eq{"type": "0"})
```
Comment (lunny) [inline models/user/user.go]: The `type` is int, the previous code is also wrong.

Comment (yp05327) [reply models/user/user.go]: Done.

File: models/user/user.go [hunk 4]
```diff
@@ -716,13 +725,25 @@ func CountUsers(ctx context.Context, opts *CountUserFilter) int64 {
...
 
-	if opts != nil && opts.LastLoginSince != nil {
-		sess = sess.Where(builder.Gte{"last_login_unix": *opts.LastLoginSince})
+	if opts != nil {
+		if opts.LastLoginSince != nil {
+			cond = cond.And(builder.Gte{"last_login_unix": *opts.LastLoginSince})
+		}
+
+		if !opts.IsAdmin.IsNone() {
+			cond = cond.And(builder.Eq{"is_admin": opts.IsAdmin})
```
Comment (lunny) [inline models/user/user.go]:
```suggestion
			cond = cond.And(builder.Eq{"is_admin": opts.IsAdmin.IsTrue()})
```

Comment (lunny) [reply models/user/user.go]: Otherwise LGTM.

**Review by lunny — APPROVED** [review]

**Pull request merged**